permutations/src/disjoint_cycles.rs

```diff
@@ -0,0 +1,261 @@
+use std::{collections::HashSet, fmt::Display, num::ParseIntError, str::FromStr};
+
+use crate::permutation_error::PermutationError;
+
+fn parse_parentheses(cycle_str: &str) -> Result<Vec<&str>, PermutationError> {
+    let mut slice_indices = Vec::new();
+    let mut start_index = 0;
+    let mut inside = false;
+    for (i, c) in cycle_str.chars().enumerate() {
+        match c {
+            '(' if inside => {
+                return Err(PermutationError::NestedCycle(cycle_str.into()));
+            }
+            ')' if !inside => {
+                return Err(PermutationError::UnmatchedParenthesis(cycle_str.into()));
+            }
+            '(' => {
+                start_index = i + 1;
+                inside = true;
+            }
+            ')' => {
+                slice_indices.push((start_index, i));
+                inside = false;
+            }
+            x if !x.is_whitespace() && !inside => {
+                return Err(PermutationError::ValueOutsideParentheses(cycle_str.into()))
+            }
+            _ => {}
+        }
+    }
+
+    // Unfinished cycle
+    if inside {
+        return Err(PermutationError::UnmatchedParenthesis(cycle_str.into()));
+    }
+
+    let slices: Vec<&str> = slice_indices
+        .into_iter()
+        .map(|(a, b)| &cycle_str[a..b])
+        .collect();
+    Ok(slices)
+}
+
+fn parse_cycle(cycle_str: &str) -> Result<Vec<usize>, PermutationError> {
+    let result: Result<Vec<usize>, ParseIntError> = cycle_str
+        .split_whitespace()
+        .map(|x| x.parse::<usize>())
+        .collect();
+
+    Ok(result?)
+}
+
+pub struct DisjointCycles<const N: usize>(pub Vec<Vec<usize>>);
```

Comment:
To help with sorting and parsing, maybe the inner vectors could be a `SortedCycle<N>` wrapper type that manages the inner invariant and implements `Ord`? then the outer one just needs to sort the outer vector at creation.